### ISSUE TITLE
tensorflow:Gradients do not exist for variables

### ISSUE THREAD

**Issue by nzplt** [yingkaisha/keras-unet-collection#7]
I try to train my dataset using oxford-iiit.ipynb; but I get this warning and it doesnt calculate validation performance

        Initial loss = 0.0
        WARNING:tensorflow:Gradients do not exist for variables ['unet3plus_down_from4_to0_0/kernel:0', 'unet3plus_down_from4_to1_0/kernel:0', 'unet3plus_down_from4_to2_0/kernel:0', 'unet3plus_down_from4_to3_0/kernel:0', 'unet3plus_down_from4_to4_0/kernel:0', 'unet3plus_down_from4_to0_0_bn/gamma:0', 'unet3plus_down_from4_to0_0_bn/beta:0', 'unet3plus_down_from4_to1_0_bn/gamma:0', 'unet3plus_down_from4_to1_0_bn/beta:0', 'unet3plus_down_from4_to2_0_bn/gamma:0', 'unet3plus_down_from4_to2_0_bn/beta:0', 'unet3plus_down_from4_to3_0_bn/gamma:0', 'unet3plus_down_from4_to3_0_bn/beta:0', 'unet3plus_down_from4_to4_0_bn/gamma:0', 'unet3plus_down_from4_to4_0_bn/beta:0', 'unet3plus_fusion_conv_4_0/kernel:0', 'unet3plus_fusion_conv_4_0_bn/gamma:0', 'unet3plus_fusion_conv_4_0_bn/beta:0', 'unet3plus_output_final/kernel:0', 'unet3plus_output_final/bias:0'] when minimizing the loss.
        WARNING:tensorflow:Gradients do not exist for variables ['unet3plus_down_from4_to0_0/kernel:0', 'unet3plus_down_from4_to1_0/kernel:0', 'unet3plus_down_from4_to2_0/kernel:0', 'unet3plus_down_from4_to3_0/kernel:0', 'unet3plus_down_from4_to4_0/kernel:0', 'unet3plus_down_from4_to0_0_bn/gamma:0', 'unet3plus_down_from4_to0_0_bn/beta:0', 'unet3plus_down_from4_to1_0_bn/gamma:0', 'unet3plus_down_from4_to1_0_bn/beta:0', 'unet3plus_down_from4_to2_0_bn/gamma:0', 'unet3plus_down_from4_to2_0_bn/beta:0', 'unet3plus_down_from4_to3_0_bn/gamma:0', 'unet3plus_down_from4_to3_0_bn/beta:0', 'unet3plus_down_from4_to4_0_bn/gamma:0', 'unet3plus_down_from4_to4_0_bn/beta:0', 'unet3plus_fusion_conv_4_0/kernel:0', 'unet3plus_fusion_conv_4_0_bn/gamma:0', 'unet3plus_fusion_conv_4_0_bn/beta:0', 'unet3plus_output_final/kernel:0', 'unet3plus_output_final/bias:0'] when minimizing the loss.
        Validation performance 0.0 is NOT improved
        Validation performance 0.0 is NOT improved
        Validation performance 0.0 is NOT improved
        Validation performance 0.0 is NOT improved
        Validation performance 0.0 is NOT improved
        Early stopping

**Comment by yingkaisha** [comment]
Hi, can you place your modified version of oxford-iiit.ipynb online? Or how you define the graph? 

The original version of this notebook does not have depth-5 layers (`down_from4_to0`). If you increased model depth, the customized model head needs some work.

**Comment by nzplt** [comment]
I put all the code, where I modify actually :) 
```
name = 'unet3plus'
  activation = 'ReLU'
  filter_num_down = [16, 32, 64, 128, 256, 512]  _#maybe its because of this ??_
  filter_num_skip = [32, 32, 32, 32, 32]
  filter_num_aggregate = 160
  
  stack_num_down = 2
  stack_num_up = 1
  n_labels = 3
  
  # `unet_3plus_2d_base` accepts an input tensor 
  # and produces output tensors from different upsampling levels
  # ---------------------------------------- #
  input_tensor = keras.layers.Input((128, 128, 3))
  # base architecture
  X_decoder = base.unet_3plus_2d_base(
      input_tensor, filter_num_down, filter_num_skip, filter_num_aggregate, 
      stack_num_down=stack_num_down, stack_num_up=stack_num_up, activation=activation, 
      batch_norm=True, pool=True, unpool=True, backbone=None, name=name)
  
  
  unet3plus = keras.models.Model([input_tensor,], OUT_stack)
  unet3plus.compile(loss=[keras.losses.categorical_crossentropy,
                          keras.losses.categorical_crossentropy,
                          keras.losses.categorical_crossentropy,
                          keras.losses.categorical_crossentropy,
                          keras.losses.categorical_crossentropy,
                          keras.losses.categorical_crossentropy],
                   loss_weights=[0.25, 0.25, 0.25, 0.25, 1.0],
                   optimizer=keras.optimizers.Adam(lr=1e-4))
  
  def input_data_process(input_array):
      '''converting pixel vales to [0, 1]'''
      return input_array/255.
  
  def target_data_process(target_array):
      '''Converting tri-mask of {1, 2, 3} to three categories.'''
      return target_array _# I dont need it, so I delete -1_
  
  valid_input = input_data_process(utils.image_to_array(image_names[ind_valid], size=128, channel=3))
  valid_target = target_data_process(utils.image_to_array(label_names[ind_valid], size=128, channel=3))
  #
  test_input = input_data_process(utils.image_to_array(image_names[ind_test], size=128, channel=3))
  test_target = target_data_process(utils.image_to_array(label_names[ind_test], size=128, channel=3))
  
  N_epoch = 100 # number of epoches
  N_batch = 16 # number of batches per epoch
  N_sample = 16 # number of samples per batch
  
  tol = 25 # current early stopping patience
  max_tol = 30 # the max-allowed early stopping patience
  min_del = 0 # the lowest acceptable loss value reduction 
  
  # loop over epoches
  for epoch in range(N_epoch):
      
      # initial loss record
      if epoch == 0:
          temp_out = unet3plus.predict([valid_input])
          y_pred = temp_out[-1]
          record = np.mean(keras.losses.categorical_crossentropy(valid_target, y_pred))
          print('\tInitial loss = {}'.format(record))
      
      # loop over batches
      for step in range(N_batch):
          # selecting smaples for the current batch
          ind_train_shuffle = utils.shuffle_ind(L_train)[:N_sample]
          
          # batch data formation
          ## augmentation is not applied
          train_input = input_data_process(
              utils.image_to_array(image_names[ind_train][ind_train_shuffle], size=128, channel=3))
          train_target = target_data_process(
              utils.image_to_array(label_names[ind_train][ind_train_shuffle], size=128, channel=3))
          
          # train on batch
          loss_ = unet3plus.train_on_batch([train_input,], [train_target, train_target, train_target, train_target, train_target, train_target])
          # ** training loss is not stored ** #
          
      # epoch-end validation
      temp_out = unet3plus.predict([valid_input])
      y_pred = temp_out[-1]
      record_temp = np.mean(keras.losses.categorical_crossentropy(valid_target, y_pred))
      # ** validation loss is not stored ** #
      
      # if loss is reduced
      if record - record_temp > min_del:
          print('Validation performance is improved from {} to {}'.format(record, record_temp))
          record = record_temp; # update the loss record
          tol = 0; # refresh early stopping patience
          # ** model checkpoint is not stored ** #
  
      # if loss not reduced
      else:
          print('Validation performance {} is NOT improved'.format(record_temp))
          tol += 1
          if tol >= max_tol:
              print('Early stopping')
              break;
          else:
              # Pass to the next epoch
              continue;
```

**Comment by yingkaisha** [comment]
`loss_weights=[0.25, 0.25, 0.25, 0.25, 1.0] # <--- pass a list of 6 numbers`

Other lines look fine to me.

**Comment by yingkaisha** [comment]
I've closed this issue since no responses were made.
